tests/basic/rwlock.rs

```diff
@@ -220,3 +221,286 @@ fn rwlock_into_inner() {
+            let reference_lock = Arc::new(std::sync::RwLock::new(()));
+
+            let threads = (0..3)
+                .map(|_| {
+                    let lock = Arc::clone(&lock);
+                    let reference_lock = Arc::clone(&reference_lock);
+                    thread::spawn(move || {
+                        for _ in 0..3 {
+                            {
+                                let _r = lock.try_read();
+                                if _r.is_ok() {
+                                    assert!(reference_lock.try_read().is_ok());
+                                }
+                            }
+                            {
+                                let _w = lock.try_write();
+                                if _w.is_ok() {
+                                    assert!(reference_lock.try_write().is_ok());
+                                }
+                            }
+                        }
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            for thd in threads {
+                thd.join().unwrap();
+            }
+        },
+        5000,
+    );
+}
+
+/// Check that any execution allowed by `std` is allowed by our `RwLock` implementation
+#[test]
+fn try_lock_implied_by_std() {
+    check_random(
+        move || {
+            let lock = Arc::new(RwLock::new(()));
+            let reference_lock = Arc::new(std::sync::RwLock::new(()));
+
+            let threads = (0..3)
+                .map(|_| {
+                    let lock = Arc::clone(&lock);
+                    let reference_lock = Arc::clone(&reference_lock);
+                    thread::spawn(move || {
+                        for _ in 0..5 {
+                            {
+                                let _r = reference_lock.try_read();
+                                if _r.is_ok() {
+                                    assert!(lock.try_read().is_ok());
+                                }
+                            }
+                            {
+                                let _w = reference_lock.try_write();
+                                if _w.is_ok() {
+                                    assert!(lock.try_write().is_ok());
+                                }
+                            }
+                        }
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            for thd in threads {
+                thd.join().unwrap();
+            }
+        },
+        5000,
+    );
+}
+
+/// Three concurrent threads, one doing an atomic increment by 1 using `write`, one trying to do an
+/// atomic increment by 1 followed by trying to do an atomic increment by 2 using `try_write`, and a
+/// third that peeks at the value using `try_read.` The `write` must succeed, while each `try_write`
+/// may or may not succeed. Thus we expect to see final values:
+/// * 0 (`try_read` ran first)
+/// * 1 (both `try_writes`s fail),
```

Comment:
But the reader may see a final value of 1 even if one (or both) try_writes were successful (if the reader went before the try_writes).  So this comment is misleading.

One thing you could do is to read the value again after the joins, and then look at all the pairs of values (value_read_by_try_read, value_read_after_joins).

Reply:
Good point. I updated the test to check all the possible pairs of values.